File: src/main/java/hacker/solution/Comparisons.java
package hacker.solution;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import static java.util.Collections.reverseOrder;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

public class Comparisons {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int testCases = Integer.parseInt(in.nextLine());

        List<_Student> studentList = new ArrayList<_Student>();
        while (testCases > 0) {
            int id = in.nextInt();
            String name = in.next();
            double cgpa = in.nextDouble();

            _Student st = new _Student(id, name, cgpa);
            studentList.add(st);

            testCases--;
        }
        studentList = studentList.stream().sorted(comparing(_Student::getCgpa, reverseOrder())
                .thenComparing(comparing(_Student::getFname))
                .thenComparing(comparing(_Student::getId)))
                .collect(toList());

        for (_Student st : studentList) {
            System.out.println(st.getFname());
        }
    }

}

File: src/main/kotlin/hacker/solutionk/Comparisons.kt
package hacker.solutionk

import java.util.*
import java.util.stream.Collectors.toList

object Comparisons {
    @JvmStatic
    fun main(args: Array<String>) {
        val `in` = Scanner(System.`in`)
        var testCases = Integer.parseInt(`in`.nextLine())

        var studentList: MutableList<Student> = ArrayList()
        while (testCases > 0) {
            val id = `in`.nextInt()
            val fname = `in`.next()
            val cgpa = `in`.nextDouble()

            val st = Student(id, fname, cgpa)
            studentList.add(st)

            testCases--
        }
//        studentList = studentList.stream()
//            .sorted(comparing<Student, Double>(Function<Student, Double> { it.getCgpa() }, reverseOrder<Double>())
//                .thenComparing(comparing<Student, String>(Function<Student, String> { it.getFname() }))
//                .thenComparing(comparing<Student, Int>(Function<Student, Int> { it.getId() }))
//            )
//            .collect<MutableList<Student>, Any>(toList<Student>())

        for (st in studentList) {
            println(st.fname)
        }
    }

}
